Source/Particles/ParticleCreation/FilterCopyTransform.H

```diff
@@ -59,32 +59,28 @@ Index filterCopyTransformParticles (DstTile& dst, SrcTile& src, Index* mask, Ind
     if (np == 0) return 0;
 
     Gpu::DeviceVector<Index> offsets(np);
-    Gpu::exclusive_scan(mask, mask+np, offsets.begin());
-
-    Index last_mask, last_offset;
-    Gpu::copyAsync(Gpu::deviceToHost, mask+np-1, mask + np, &last_mask);
-    Gpu::copyAsync(Gpu::deviceToHost, offsets.data()+np-1, offsets.data()+np, &last_offset);
-
-    Gpu::streamSynchronize();
-    const Index num_added = N * (last_mask + last_offset);
+    auto total = amrex::Scan::ExclusiveSum(np, mask, offsets.data());
+    const Index num_added = N * total;
     dst.resize(std::max(dst_index + num_added, dst.numParticles()));
 
     const auto p_offsets = offsets.dataPtr();
 
     const auto src_data = src.getParticleTileData();
     const auto dst_data = dst.getParticleTileData();
 
-    AMREX_HOST_DEVICE_FOR_1D( np, i,
+    amrex::ParallelForRNG(np,
+    [=] AMREX_GPU_DEVICE (int i, amrex::RandomEngine const& engine) noexcept
     {
         if (mask[i])
         {
-            for (int j = 0; j < N; ++j)
-                copy(dst_data, src_data, i, N*p_offsets[i] + dst_index + j);
+            for (int j = 0; j < N; ++j) {
+                copy(dst_data, src_data, i, N*p_offsets[i] + dst_index + j, engine);
+            }
             transform(dst_data, src_data, i, N*p_offsets[i] + dst_index);
         }
-    })
+    });
 
-    Gpu::streamSynchronize();
+    Gpu::synchronize();
```

Comment:
@WeiqunZhang Could you explain why we need `synchronize` rather than `streamSynchronize` here?

Reply:
Because we no longer use locks, we do not want to have multiple ParallelForRNG running.  (Note that previously even without the local temporary DeviceVector variable, we would still need to sync because the lock does not work with multiple streams either.)  So ParallelForRNG is not running on the null stream.  `Gpu::streamSynchronize` only synchronizes for the current stream, which might not be the null stream.  (Also note that in this case before the change, `Gpu::streamSynchronize` is essentially the same as  `Gpu::synchronize` because it cannot return before `ParallelFor. finishes and therefore the next stream will be not used after this stream finishes.)